### PR TITLE
[207] Topic model optimisation

### LINKED ISSUE #207: Add method for finding optimal number of topics

### PR DESCRIPTION
**Pull request by Jack-0-0**
Closes #207.

This PR adds:

- Functions to `analysis/lda_modelling_utils.py` to find optimal number of iterations and clusters for a topic model.
- Notebook showing usage of above functions (`innovation_sweet_spots/analysis/examples/12_tomotopy_topic_optimisation_gtr.py`
- Readme with instructions on how to install tomotopy on mac m1 (`analysis/m1_mac_tomotopy_installation.md`)
 

---

Checklist:

- [x] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [x] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have merged any new changes from `dev`
- [x] I have documented the code
  - [x] Major functions have docstrings
  - [x] Appropriate information has been added to `README`s
- [x] I have explained the feature in this PR or (better) in `output/reports/`
- [x] I have requested a code review

### CODE REVIEW

**Review by beingkk — CHANGES_REQUESTED** [review]
Thanks so much @Jack-0-0 ! 

It looks very good to me - just couple quick comments, let me know if we could implement them as part of this PR:

- Could we also add another simple function to `lda_modelling_utils` plot that shows the coherence score (vertical axis) versus the number of topics (horizontal axis)
- Do you think we could, similarly to my comment on the clustering PR, run the topic modelling multiple times for each number of topics  - and hence obtain robust values for coherence? I appreciate it will take a long time to run that for the user, but it only needs to be done once - and can also be in theory parallelised easly.

**Comment by Jack-0-0** [comment]
> Thanks so much @Jack-0-0 !
>
Thanks for reviewing!
> It looks very good to me - just couple quick comments, let me know if we could implement them as part of this PR:
> 
> * Could we also add another simple function to `lda_modelling_utils` plot that shows the coherence score (vertical axis) versus the number of topics (horizontal axis)

I'm unsure how this plot would look, what should I do with iterations? The current plot shows both coherence score and iterations versus coherence score.

> * Do you think we could, similarly to my comment on the clustering PR, run the topic modelling multiple times for each number of topics  - and hence obtain robust values for coherence? I appreciate it will take a long time to run that for the user, but it only needs to be done once - and can also be in theory parallelised easly.

I have added this. The function `topic_model_param_search` now has a parameter `random_seeds`. The function will loop through all the random seeds provided and then average the coherence score values across the different runs.

File: innovation_sweet_spots/analysis/lda_modelling_utils.py [hunk 1]
```diff
@@ -337,6 +335,90 @@ def make_pyLDAvis(topic_model: LDAModel, output_filepath: PathLike):
...
+    """
+    iteration_record = []
+    n_topics_record = []
+    coherence_score_record = []
+    random_seed_record = []
+    for random_seed in tqdm(random_seeds):
+        for k in tqdm(n_topics):
+            model = tp.LDAModel(k=k, corpus=corpus, min_df=1, seed=random_seed)
+            for i in tqdm(
+                range(0, max_iterations + n_iters_logging, n_iters_logging),
```
Comment (beingkk) [inline innovation_sweet_spots/analysis/lda_modelling_utils.py]:
```suggestion
                range(0, max_iterations, n_iters_logging),
```

I think if you have `max_iterations + n_iters_logging` it will run the training for one cycle longer (cycle = `n_iters_logging`). When you plot the results, it starts with "0" iterations, but that's actually already `n_iters_logging` number of iterations.

File: innovation_sweet_spots/analysis/lda_modelling_utils.py [hunk 2]
```diff
@@ -337,6 +335,90 @@ def make_pyLDAvis(topic_model: LDAModel, output_filepath: PathLike):
...
+                desc=f"{k} topics iterations",
+            ):
+                model.train(n_iters_logging)
+                coherence_score = tp.coherence.Coherence(
+                    model, coherence="c_v"
+                ).get_score()
+                iteration_record.append(i)
```
Comment (beingkk) [inline innovation_sweet_spots/analysis/lda_modelling_utils.py]:
```suggestion
                iteration_record.append(i + n_iters_logging)
```
I think in this way you'll keep a correct log of the number of iterations

Comment (Jack-0-0) [reply innovation_sweet_spots/analysis/lda_modelling_utils.py]: Good spot, thanks. Your suggestion worked, I've made the changes 🙂

**Comment by beingkk** [comment]
Thanks a lot!
